sdks/java/extensions/histogram/src/main/java/org/apache/beam/sdk/extensions/histogram/Histogram.java

```diff
@@ -0,0 +1,571 @@
+    }
+
+    @Override
+    public HistogramAccumulator mergeAccumulators(
+        @UnknownKeyFor @NonNull @Initialized Iterable<HistogramAccumulator> accumulators) {
+      HistogramAccumulator merged = createAccumulator();
+
+      for (HistogramAccumulator histogramAccumulator : accumulators) {
+        checkArgument(
+            merged.counts.length == histogramAccumulator.counts.length,
+            "number of buckets in the merging accumulators should be the same.");
+        for (int i = 0; i < histogramAccumulator.counts.length; ++i) {
+          merged.counts[i] += histogramAccumulator.counts[i];
+        }
+      }
+      return merged;
+    }
+
+    @Override
+    public List<Long> extractOutput(HistogramAccumulator accumulator) throws NullPointerException {
+      checkNotNull(accumulator, "can not output from null histogram.");
+      return Longs.asList(accumulator.counts);
+    }
+
+    @Override
+    public @UnknownKeyFor @NonNull @Initialized Coder<HistogramAccumulator> getAccumulatorCoder(
+        @UnknownKeyFor @NonNull @Initialized CoderRegistry registry,
+        @UnknownKeyFor @NonNull @Initialized Coder<T> inputCoder) {
+      return new HistogramAccumulatorCoder();
+    }
+
+    @Override
+    public @UnknownKeyFor @NonNull @Initialized Coder<List<Long>> getDefaultOutputCoder(
+        @UnknownKeyFor @NonNull @Initialized CoderRegistry registry,
+        @UnknownKeyFor @NonNull @Initialized Coder<T> inputCoder) {
+      return ListCoder.of(VarLongCoder.of());
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.add(DisplayData.item("numBuckets", bounds.length + 1).withLabel("Number of buckets"));
+    }
+  }
+
+  /** Accumulator of the Histogram combiner. */
+  static final class HistogramAccumulator {
+
+    private long[] counts;
+
+    public HistogramAccumulator(int numBuckets) {
+      checkArgument(
+          numBuckets > 2,
+          "number of buckets should be greater than two - underflow bucket and overflow bucket.");
+      this.counts = new long[numBuckets];
+    }
+
+    @Override
+    public boolean equals(@Nullable Object object) {
+      if (object instanceof HistogramAccumulator) {
+        HistogramAccumulator other = (HistogramAccumulator) object;
+        return Arrays.equals(counts, other.counts);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(Arrays.hashCode(counts));
+    }
+  }
+
+  /** Coder for {@link HistogramAccumulator}. */
+  static final class HistogramAccumulatorCoder extends CustomCoder<HistogramAccumulator> {
+
+    private static final VarLongCoder LONG_CODER = VarLongCoder.of();
+    private static final VarIntCoder INT_CODER = VarIntCoder.of();
+
+    /**
+     * Index to indicate method used where only non-empty buckets are encoded with their indices.
+     */
+    private static final int NON_EMPTY_BUCKETS_CODER = 0;
```

Comment:
we can still encode the enum as an integer, my suggestion was mostly "cosmetic" so that the code is more clear and the constants have some meaning attached to them by being part of a custom named type (an enum) instead of being just an int

Reply:
done